prow/commentpruner/commentpruner.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// commentpruner facilitates efficiently deleting issue comments as a reaction to webhook events.
+package commentpruner
+
+import (
+	"sync"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+)
+
+// bufSize is the size of the sliding window and subsequently the cache size.
+const bufSize = 100
+
+type githubClient interface {
+	ListIssueComments(org, repo string, number int) ([]github.IssueComment, error)
+	DeleteComment(org, repo string, id int) error
+}
+
+// EventClient is a struct that provides comment deletion for an event related to an issue.
+// A single client instance should be created for each event and shared by all consumers of the event.
+// The client fetches the comments only once and filters that list repeatedly to find comments to
+// delete. This avoids using lots of API tokens when fetching comments for each handler that wants
+// to delete comments. (An HTTP cache only partially helps with this because deletions modify the
+// list of comments so the next call requires GH to send the resource again.)
+type EventClient struct {
+	org    string
+	repo   string
+	number int
+
+	ghc githubClient
+	log *logrus.Entry
+
+	comments []github.IssueComment
+
+	once sync.Once  // Used to fetch the IssueComments once.
+	lock sync.Mutex // Protects `comments`.
```

Comment:
As a convention locks are added on top of the fields they protect.

Reply:
That makes sense. Fixed.